include/nbl/builtin/hlsl/scan/default_scheduler.hlsl

```diff
@@ -35,187 +17,181 @@ namespace scan
-				_scanParams.temporaryStorageOffset[1] = _scanParams.temporaryStorageOffset[0]+WorkgroupCount(0);
-				_scanParams.temporaryStorageOffset[2] = _scanParams.temporaryStorageOffset[1]+WorkgroupCount(1);
-				break;
-			default:
-				break;
-#if NBL_BUILTIN_MAX_SCAN_LEVELS>7
+        const uint32_t workgroupSizeLog2 = firstbithigh(glsl::gl_WorkGroupSize().x);
+        _scanParams.lastElement[0] = elementCount-1u;
+        _scanParams.topLevel = firstbithigh(_scanParams.lastElement[0]) / workgroupSizeLog2;
+        
+        for (uint32_t i=0; i<NBL_BUILTIN_MAX_LEVELS/2;)
+        {
+            const uint32_t next = i+1;
+            _scanParams.lastElement[next] = _scanParams.lastElement[i]>>workgroupSizeLog2;
+            i = next;
+        }
+        _schedulerParams.cumulativeWorkgroupCount[0] = WorkgroupCount(0);
+        _schedulerParams.finishedFlagOffset[0] = 0u;
+        switch(_scanParams.topLevel)
+        {
+            case 1u:
+                _schedulerParams.cumulativeWorkgroupCount[1] = _schedulerParams.cumulativeWorkgroupCount[0]+1u;
+                _schedulerParams.cumulativeWorkgroupCount[2] = _schedulerParams.cumulativeWorkgroupCount[1]+WorkgroupCount(0);
+                // climb up
+                _schedulerParams.finishedFlagOffset[1] = 1u;
+                
+                _scanParams.temporaryStorageOffset[0] = 2u;
+                break;
+            case 2u:
+                _schedulerParams.cumulativeWorkgroupCount[1] = _schedulerParams.cumulativeWorkgroupCount[0]+WorkgroupCount(1);
+                _schedulerParams.cumulativeWorkgroupCount[2] = _schedulerParams.cumulativeWorkgroupCount[1]+1u;
+                _schedulerParams.cumulativeWorkgroupCount[3] = _schedulerParams.cumulativeWorkgroupCount[2]+WorkgroupCount(1);
+                _schedulerParams.cumulativeWorkgroupCount[4] = _schedulerParams.cumulativeWorkgroupCount[3]+WorkgroupCount(0);
+                // climb up
+                _schedulerParams.finishedFlagOffset[1] = WorkgroupCount(1);
+                _schedulerParams.finishedFlagOffset[2] = _schedulerParams.finishedFlagOffset[1]+1u;
+                // climb down
+                _schedulerParams.finishedFlagOffset[3] = _schedulerParams.finishedFlagOffset[1]+2u;
+                
+                _scanParams.temporaryStorageOffset[0] = _schedulerParams.finishedFlagOffset[3]+WorkgroupCount(1);
+                _scanParams.temporaryStorageOffset[1] = _scanParams.temporaryStorageOffset[0]+WorkgroupCount(0);
+                break;
+            case 3u:
+                _schedulerParams.cumulativeWorkgroupCount[1] = _schedulerParams.cumulativeWorkgroupCount[0]+WorkgroupCount(1);
+                _schedulerParams.cumulativeWorkgroupCount[2] = _schedulerParams.cumulativeWorkgroupCount[1]+WorkgroupCount(2);
+                _schedulerParams.cumulativeWorkgroupCount[3] = _schedulerParams.cumulativeWorkgroupCount[2]+1u;
```

Comment:
because due to forward progress guarantees you can only do upsweep or downsweep in a single dispatch, you don't need the "climb down"

you probably don't even need the cumulative workgroup counts, as the way you'd find out if your current workgroup truly is the last one, is by checking the return value of the finished flag atomic counter.

e.g. `markFinished`
```cpp
uint32_t howManyDone = flagsAccessor.atomicAdd(finishedOffset[level]+virtualWorkgroupIndexInThisLevel);
//uint32_t currentWorkgroupCount = currentElementCount>>ItemsPerWorkgroupLog2;
uint32_t nextLevelWorkgroupCount = currentWorkgroupCount>>ItemsPerWorkgroupLog2;
// NOT: either last workgroup, or in the last bundle of workgroups and last workgroup dynamically there
if (howManyDone!=(ItemsPerWorkgroup-1) && ((virtualWorkgroupIndexInThisLevel>>ItemsPerWorkgroupLog2)!=nextLevelWorkgroupCount || (virtualWorkgroupIndexInThisLevel&(ItemsPerWorkgroupLog2-1))!=howManyDone) )
    return;
// last one takes over
currentWorkgroupCount = nextLevelWorkgroupCount;
```

Reply:
btw I'd try to rewrite this as a loop instead of a weird switch that was only written like this to avoid recursion in GLSL